src/datasets/arrow_dataset.py

```diff
@@ -650,13 +647,24 @@ def flatten_(self, max_depth=16):
             "Flattened dataset from depth {} to depth {}.".format(depth, 1 if depth + 1 < max_depth else "unknown")
         )
 
+    @fingerprint_transform(inplace=False)
+    def flatten(self, new_fingerprint, max_depth=16) -> "Dataset":
+        """Flattens the table.
+        Each column with a struct type is flattened into one column per struct field.
+        Other columns are left unchanged.
+
+        Returns:
+            A copy of the dataset with flattened columns
+        """
+        dataset = copy.deepcopy(self)
+        dataset._fingerprint = new_fingerprint
+        dataset.flatten_(max_depth=max_depth)
```

Comment:
flatten_ already updates the fingerprint so we're updating the fingerprint twice here.
We could simply copy paste the code from flatten_. I think it's ok since we may deprecate flatten_ at one point, and the code is short and straightforward.

This also applies for the other transforms. Let me know what you think

Reply:
I also added a `@deprecated` decorator that emits a DeprecationWarning when calling the methods

Reply:
Nice thanks !
Although usually deprecation warnings are only emitted once, maybe we can just give an id parameter to the deprecated decorator so that the second time a deprecated function is called we can say that the warning has already been emitted and therefore keep it silent ? We can use a dictionary to keep track of already emitted deprecation warning as in `transformers`.

Reply:
Sounds good!